python/lsst/rubintv/production/performance.py

```diff
@@ -786,6 +906,373 @@ def callback(self, payload: Payload) -> None:
+    """
+    if not timings:
+        axTop.set_axis_off()
+        return
+
+    items: list[tuple[str, float]] = sorted(timings.items(), key=lambda kv: kv[1])
+    names: list[str] = [k for k, _ in items]
+    times: list[float] = [v for _, v in items]
+    n = len(times)
+
+    heights = np.linspace(yMax, yMin, n, dtype=float)
+
+    # top: staircase heights; bottom: full-height guides
+    for i, t in enumerate(times):
+        axTop.vlines(t, 0.0, float(heights[i]), linestyles="--", linewidth=1.2)
+        axBottom.axvline(t, color="black", linestyle="--", linewidth=1.2, ymin=0, ymax=1)
+
+    axTop.text(
+        times[0],
+        1,
+        f"{names[0]} (+{times[0]:.2f}s)",
+        rotation=45,
+        rotation_mode="anchor",
+        ha="left",
+        va="bottom",
+    )
+
+    # arrows + labels between consecutive events
+    labelDy = 0.03 * (yMax - yMin)
+    for i in range(n - 1):
+        t0, t1 = times[i], times[i + 1]
+        yRight = float(heights[i + 1])
+        dt = t1 - t0
+
+        # Horizontal arrow pointing to the top of the right event line
+        axTop.annotate(
+            "",
+            xy=(t1, yRight),
+            xytext=(t0, yRight),
+            arrowprops=dict(arrowstyle="<->", linewidth=1.2),
+        )
+
+        # Place the label above the right event point (not at the midpoint)
+        axTop.text(
+            t1,
+            yRight + labelDy,
+            f"{names[i + 1]} (+{dt:.2f}s)",
+            rotation=45,
+            rotation_mode="anchor",
+            ha="left",
+            va="bottom",
+        )
+
+    axTop.set_ylim(0.0, yMax + 2 * labelDy)
+    axTop.set_yticks([])
+    axTop.set_ylabel("Events", labelpad=6)
+    axTop.grid(False)
+
+    # Remove top x-axis and side axis lines
+    for spine in ("top", "left", "right"):
+        axTop.spines[spine].set_visible(False)
+    axTop.tick_params(axis="x", top=False)
+
+
+def createLegendBoxes(
+    axTop: Axes,
+    colors: dict[str, str],
+    extraLines: list[str] | None = None,
+) -> None:
+    """Two axis-anchored legends at the bottom-right of *axTop*.
+
+    Left block = colored task entries; right block = free-text lines.
+    Both are placed relative to axTop's axes coordinates (0..1).
+    """
+    # Left: colored task entries (placed just to the *left* of the text legend)
+    colorHandles = [Patch(facecolor=v, label=k) for k, v in colors.items()]
+    tasksLegend = axTop.legend(
+        handles=colorHandles,
+        loc="lower right",
+        bbox_to_anchor=(1.0, 0.0),  # bottom-right corner of axTop
+        bbox_transform=axTop.transAxes,
+        frameon=False,
+        ncol=1,
+        borderaxespad=0.0,
+        fontsize="medium",
+    )
+
+    if extraLines:
+        # Force a draw so we can compute the first legend's bbox.
+        fig = axTop.figure
+        fig.canvas.draw()
+        # type ignore because get_renderer() only exists with Agg backend
```

Comment:
Don't need to change, and I didn't spend any time looking into how you're ensuring you have an Agg backend, but searching the stack,  I think we've dealt with stuff like this like 

```
from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
canvas = FigureCanvas(fig)
canvas.draw()
renderer = canvas.get_renderer()
```

in https://github.com/lsst/fgcm/blob/8adec12ab3658fce58a7546ba4c50ee0ce1bdb9b/fgcm/fgcmUtilities.py#L1032